File: demo/KSYLiveDemo/KSYDemoUI/KSYStateLableView.m
//
//  KSYStateLableView.m
//  KSYDemo
//
//  Created by pengbin on 16/9/5.
//  Copyright © 2016年 ksyun. All rights reserved.
//

#import "KSYUIVC.h"
#import "KSYStateLableView.h"
#import "KSYPresetCfgView.h"


@interface KSYStateLableView ()

@end

@implementation KSYStateLableView

- (id) init {
    self = [super init];
    self.backgroundColor = [UIColor clearColor];
    self.textColor = [UIColor redColor];
    self.numberOfLines = 7;
    self.textAlignment = NSTextAlignmentLeft;
    [self initStreamStat];
    return self;
}

// 将推流状态信息清0
- (void) initStreamStat{
    memset(&_lastStD, 0, sizeof(_lastStD));
    _startTime  = [[NSDate date]timeIntervalSince1970];
    _notGoodCnt = 0;
    _bwRaiseCnt = 0;
    _bwDropCnt  = 0;
}

- (void) updateState:(KSYStreamerBase*)str {
    StreamState curState = {0};
    curState.timeSecond     = [[NSDate date]timeIntervalSince1970];
    curState.uploadKByte    = [str uploadedKByte];
    curState.encodedFrames  = [str encodedFrames];
    curState.droppedVFrames = [str droppedVideoFrames];
    StreamState deltaS  = {0};
    deltaS.timeSecond    = curState.timeSecond    -_lastStD.timeSecond    ;
    deltaS.uploadKByte   = curState.uploadKByte   -_lastStD.uploadKByte   ;
    deltaS.encodedFrames = curState.encodedFrames -_lastStD.encodedFrames ;
    deltaS.droppedVFrames= curState.droppedVFrames-_lastStD.droppedVFrames;
    _lastStD = curState;
    
    double realTKbps   = deltaS.uploadKByte*8 / deltaS.timeSecond;
    double encFps      = deltaS.encodedFrames / deltaS.timeSecond;
    double dropPercent = deltaS.droppedVFrames * 100.0 /MAX(curState.encodedFrames, 1);
    
    NSString* liveTime =[KSYUIVC timeFormatted: (int)(curState.timeSecond-_startTime) ] ;
    NSString *uploadDateSize = [KSYUIVC sizeFormatted:curState.uploadKByte];
    NSString* stateurl  = [NSString stringWithFormat:@"%@\n", [str.hostURL absoluteString]];
    NSString* statekbps = [NSString stringWithFormat:@"实时码率(kbps)%4.1f\tA%4.1f\tV%4.1f\n", realTKbps, [str encodeAKbps], [str encodeVKbps] ];
    NSString* statefps  = [NSString stringWithFormat:@"实时帧率(fps)%2.1f\t总上传:%@\n", encFps, uploadDateSize ];
    NSString* statedrop = [NSString stringWithFormat:@"视频丢帧 %4d\t %2.1f%% \n", curState.droppedVFrames, dropPercent ];
    NSString* netEvent = [NSString stringWithFormat:@"网络事件计数 %d bad\t bw %d Raise\t %d drop\n", _notGoodCnt, _bwRaiseCnt, _bwDropCnt];
    NSString *cpu_use = [NSString stringWithFormat:@"cpu: %3.2f \t%@",[KSYUIVC cpu_usage], liveTime];
    
    self.text = [ stateurl   stringByAppendingString:statekbps ];
    self.text = [ self.text  stringByAppendingString:statefps  ];
    self.text = [ self.text  stringByAppendingString:statedrop ];
    self.text = [ self.text  stringByAppendingString:netEvent  ];
    self.text = [ self.text  stringByAppendingString:cpu_use  ];
    
}

- (void)drawTextInRect:(CGRect) rect {
    if (self.text == nil){
        return;
    }
    CGFloat oldH = rect.size.height;
    NSAttributedString *attributedText;
    attributedText = [[NSAttributedString alloc] initWithString:self.text
                                                     attributes:@{NSFontAttributeName:self.font}];
    rect.size.height = [attributedText boundingRectWithSize:rect.size
                                                    options:NSStringDrawingUsesLineFragmentOrigin
                                                    context:nil].size.height;
    if (self.numberOfLines != 0) {
        rect.size.height = MIN(rect.size.height, self.numberOfLines * self.font.lineHeight);
    }
    rect.origin.y = oldH - rect.size.height;  // 底部对齐 将一段文字移动到最底部
    [super drawTextInRect:rect];
}
@end

File: demo/KSYLiveDemo_Swift/KSYDemoUI_Swift/subviews/KSYStateLableView.swift
//
//  KSYStateLableView.swift
//  KSYLiveDemo_Swift
//
//  Created by iVermisseDich on 17/1/10.
//  Copyright © 2017年 com.ksyun. All rights reserved.
//

import UIKit

struct StreamState {
    var timeSecond: Double;     // 更新时间
    var uploadKByte: Int;       // 上传的字节数(KB)
    var encodedFrames: Int;     // 编码的视频帧数
    var droppedVFrames: Int;    // 丢弃的视频帧数
}


/**
 KSY 推流SDK的状态监控控件
 
 streamerVC 每秒钟刷新一次监控的数据
 视图中文字为底部对齐
 */


class KSYStateLableView: UILabel {
    // 上一次更新时的数据, 假定每秒更新一次
    var _lastStD: StreamState?
    
    var startTime: Double? = 0      // 开始推流的时间
    var notGoodCnt: Int?            // 网络拥塞事件发生次数
    var bwRaiseCnt: Int?            // 码率上调事件发生次数
    var bwDropCnt: Int?             // 码率下调事件发生次数
    
    init() {
        super.init(frame: CGRect.zero)
        backgroundColor = .clear
        textColor = .red
        numberOfLines = 7
        textAlignment = .left
        initStreamStat()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // 将推流状态信息清0
    func initStreamStat() {
        // TODO: 不确定是否写对了
        memset(&_lastStD, 0, MemoryLayout<StreamState>.size)
        startTime = Date().timeIntervalSince1970
        notGoodCnt = 0
        bwRaiseCnt = 0
        bwDropCnt = 0
    }
    
    // 更新数据(需要每秒被调用一次)
    func updateState(str: KSYStreamerBase) {
        var curState: StreamState = StreamState(timeSecond: 0,uploadKByte: 0,encodedFrames: 0,droppedVFrames: 0)
        curState.timeSecond = Date().timeIntervalSince1970
        curState.uploadKByte = Int(str.uploadedKByte)
        curState.encodedFrames = Int(str.encodedFrames)
        curState.droppedVFrames = Int(str.droppedVideoFrames)
        
        var deltaS: StreamState = StreamState(timeSecond: 0,uploadKByte: 0,encodedFrames: 0,droppedVFrames: 0)
        if let _ = _lastStD {
        }else{
            deltaS = curState
        }
        
        if _lastStD == nil {
            _lastStD = StreamState(timeSecond: 0, uploadKByte: 0, encodedFrames: 0, droppedVFrames: 0)
        }
        
        deltaS.timeSecond = curState.timeSecond - _lastStD!.timeSecond
        deltaS.uploadKByte = curState.uploadKByte - _lastStD!.uploadKByte
        deltaS.encodedFrames = curState.encodedFrames - _lastStD!.encodedFrames
        deltaS.droppedVFrames = curState.droppedVFrames - _lastStD!.droppedVFrames
        _lastStD = curState
        
        let realTKbps = Double(deltaS.uploadKByte * 8) / deltaS.timeSecond
        let encFps = Double(deltaS.encodedFrames) / deltaS.timeSecond
        let dropPercent = deltaS.droppedVFrames * 100 / max(curState.encodedFrames, 1)
        
        let liveTime = KSYUIVC().timeFormatted(totalSeconds: Int(curState.timeSecond - startTime!))
        let uploadDateSize = KSYUIVC().sizeFormatted(kb: curState.uploadKByte)
        let stateUrl = "\(str.hostURL.absoluteString)\n"
        let stateKbps = String.init(format: "实时码率(kbps)%4.1f\tA%4.1f\tV%4.1f\n", realTKbps, str.encodeAKbps, str.encodeVKbps)
        let stateFps = String.init(format: "实时帧率(fps)%2.1f\t总上传:%@\n", encFps, uploadDateSize )
        let stateDrop = String.init(format: "视频丢帧 %4d\t %2.1f%% \n", curState.droppedVFrames, dropPercent)
        let netEvent = String.init(format: "网络事件计数 %d bad\t bw %d Raise\t %d drop\n", notGoodCnt!, bwRaiseCnt!, bwDropCnt!)
        let cup_use = String.init(format: "%@ \tcpu: %.2f mem: %.1fMB",liveTime, KSYHelper.cpu_usage(), KSYHelper.memory_usage())
        
        if let _ = text {
        }else{
            text = ""
        }
        text = stateUrl + stateKbps
        text! += stateFps
        text! += stateDrop
        text! += netEvent
        text! += cup_use
    }
    
    override func drawText(in rect: CGRect) {
        guard let _ = text else { return }
        
        var newRect = rect
        let oldH: CGFloat = newRect.height
        
        let attributeText = NSAttributedString.init(string: text!,
                                                    attributes: [NSFontAttributeName:self.font])
        
        
        newRect.size = CGSize.init(width: newRect.width, height: attributeText.boundingRect(with: newRect.size, options: .usesLineFragmentOrigin, context: nil).height)
        
        if numberOfLines != 0 {
            newRect.size = CGSize.init(width: newRect.width, height: min(newRect.height, CGFloat(numberOfLines) * font.lineHeight))
        }
        newRect.origin.y = oldH - newRect.height
        super.drawText(in: newRect)
    }
}
